app/src/main/java/com/asfoundation/wallet/viewmodel/TransactionsViewModel.java

```
@@ -96,32 +89,89 @@
     this.viewScheduler = viewScheduler;
     this.networkScheduler = networkScheduler;
     this.disposables = new CompositeDisposable();
+    init();
   }
 
-  @Override protected void onCleared() {
-    super.onCleared();
-    hasTransactions = false;
-    if (!disposables.isDisposed()) {
-      disposables.dispose();
-    }
-    handler.removeCallbacks(startFetchTransactionsTask);
-    handler.removeCallbacks(startGlobalBalanceTask);
+  private void init() {
+    progress.postValue(true);
+    handleBalanceWalletsExperiment();
+    handleTopUpClicks();
+    handleUnreadConversationCount();
+    handlePromotionTooltipVisibility();
+    handlePromotionUpdateNotification();
+    handleRateUsDialogVisibility();
+    handleConversationCount();
+    handleWalletData();
+    handleFingerprintTooltipVisibility();
+  }
+
+  public void updateData() {
+    refreshData.onNext(true);
+  }
+
+  public void stopRefreshingData() {
+    refreshData.onNext(false);
+  }
+
+  /**
+   * Spine stream responsible for keeping wallet data up-to-date.
+   *
+   * If {@link #refreshData} is true, we retrieve the latest network and active wallet, we post
+   * its values, we check if the wallet changed and refresh the rest of the data.
+   *
+   * We use switchMap to disregard previous subscriptions every time we need to refresh data.
+   *
+   * @see #refreshTransactionsAndBalance(TransactionsWalletModel model) .
+   */
+  private void handleWalletData() {
+    disposables.add(observeRefreshData().switchMap(__ -> observeNetworkAndWallet())
+        .doOnNext(walletNetworkModel -> {
+          if (walletNetworkModel.isNewWallet()) {
+            defaultWalletModel.postValue(walletNetworkModel);
+          }
+        })
+        .switchMapCompletableDelayError(this::updateWalletData)
```

Errors are not being handled as empty list state.